contracts/ProjectValidation.sol

```diff
@@ -0,0 +1,180 @@
+    ) {
+        starter = msg.sender;
+        manager = _manager;
+        checker = Checker(
+            _checker,
+            false,
+            false,
+            false
+        );
+        exchangerContract = _exchangerContract;
+        fundTokens = SimpleToken(fundTokenAddress);
+        workers = _workers;
+        signatures[starter] = false;
+        signatures[manager] = false;
+        for (uint8 i = 0; i < workers.length; i++) {
+            workersBalances[workers[i]] = 0;
+        }
+    }
+
+    modifier onlyValidator() {
+        require(msg.sender == manager || msg.sender == starter || msg.sender == checker.addr);
+        _;
+    }
+
+    modifier onlyChecker(){
+        require(msg.sender == checker.addr && checker.presence);
+        _;
+    }
+
+    modifier atStage(Stages _stage) {
+        require(stage == _stage);
+        _;
+    }
+
+    function changeStateTo (Stages _stage) internal {
+        StateChanged(stage, _stage);
+        stage = _stage;
+    }
+
+    modifier afterExecutingGoToState(Stages _stage){
+        _;
+        changeStateTo(_stage);
+    }
+
+    function sign() onlyValidator atStage(Stages.collectingSignatures) returns (bool signed) {
+        signatures[msg.sender] = true;
+        return true;
+    }
+
+    function checkerSign(bool signature) onlyChecker atStage(Stages.checkerWork) afterExecutingGoToState(Stages.readyToComplete) {
+        checker.signed = true;
+        checker.signature = signature;
+    }
+
+    function checkerCall() internal {
+        changeStateTo(Stages.checkerWork);
+        checker.presence = true;
+    }
+
+    function getWorkerBalance(address worker) returns (uint balance) {
+        uint fundWorkerBalance = workersBalances[worker];
+        balance = fundWorkerBalance * 2 / 3 * additionalTokenRate;
+    }
+
+    function tryToCompleteProject() onlyValidator {
+        if ( ( !signatures[starter] || !signatures[manager] ) && !checker.presence) {
+            checkerCall();
+        } else if ( (checker.signed && checker.signature) || (signatures[starter] && signatures[manager]) ) {
+            changeStateTo(Stages.readyToComplete);
+            completeProject(true);
+        } else {
+            changeStateTo(Stages.readyToComplete);
+            completeProject(false);
+        }
+    }
+
+    function completeProject(bool decision) internal atStage(Stages.readyToComplete) afterExecutingGoToState(Stages.closing) {
+        projectCompleted = decision;
+        projectBalance = fundTokens.balanceOf(this);
+    }
+
+    function closeProject() onlyValidator atStage(Stages.closing) {
+        projectClosed = true;
+        if (projectCompleted) {
+            changeStateTo(Stages.successfullyClosed);
+        }else {
+            changeStateTo(Stages.unsuccessfullyClosed);
+        }
+    }
+
+    function sendTokensToWorkers() atStage(Stages.successfullyClosed) returns (uint amount) {
+        for (uint8 i = 0; i < workers.length; i++) {
```

Comment:
This loop can stuck because of gas limit, if you have too many workers
In order to prevent this, you can pass `start` and `end` as a parameter, where `i = start; i <= end`

Reply:
thanks, I already fixed this